tests/autorestart_test/test_container_autorestart.py

```diff
@@ -0,0 +1,256 @@
+            return process_info.split()[1].strip()
+
+def get_process_info(duthost, container_name, process_name):
+    """
+    @summary: Get the pid of process in the specified container by analyzing
+              the command output of "ps -ax"
+    @return: first value is a boolean. True indicates the processs is running
+             second value is the process id. Returning -1 if the process is not found
+    """
+    running_status = False
+    process_id = -1
+ 
+    ps_cmd_out = duthost.shell("docker exec {} ps -ax".format(container_name))
+    for line in ps_cmd_out["stdout_lines"]:
+        if "/usr/bin/supervisor-proc-exit-listener" not in line \
+            and line.find(process_name) != -1:
+            running_status = True
+            process_id = int(line.split()[0].strip())
+            break 
+
+    return running_status, process_id
+ 
+def kill_process_by_name(duthost, container_name, process_name):
+    """
+    @summary: Kill a process in the specified container by its name
+
+    """
+    running_status, process_id = get_process_info(duthost, container_name, process_name)
+    if running_status:
+        duthost.shell("docker exec {} kill -SIGKILL {}".format(container_name, process_id)) 
+    else:
+        pytest_assert(False, "Failed to find {} process in {}".format(process_name, container_name))
+
+    time.sleep(7)
+
+    is_running = is_container_running(duthost, container_name)
+    if is_running == "true":
+        running_status, process_id = get_process_info(duthost, container_name, process_name)
+        if running_status:
+            pytest_assert(False, "Failed to stop {} process before test".format(process_name))
+
+    logging.info("{} process in {} is stopped successfully".format(process_name, container_name)) 
+
+def kill_process_by_pid(duthost, container_name, program_name, program_pid):
+    """
+    @summary: Kill a process in the specified container by its pid
+
+    """
+    duthost.shell("docker exec {} kill -SIGKILL {}".format(container_name, program_pid)) 
+
+    time.sleep(7)
+
+    is_running = is_container_running(duthost, container_name)
+    if is_running == "true":
+        running_status = get_program_state(duthost, container_name, program_name)
+        if running_status == "RUNNING":
+            pytest_assert(False, "Failed to stop {} before test".format(program_name))
+
+    logging.info("{} in {} is stopped successfully".format(program_name, container_name)) 
+
+
+def check_container_status(duthost, container_name, should_be_stopped):
+    """
+    @summary: Determine whether a container should be in running state or not
+    """
+    is_running = is_container_running(duthost, container_name)
+    if is_running == "false" and should_be_stopped:
+        return True
+    if is_running == "true" and not should_be_stopped:
+        return True
+    return False
+ 
+def verify_autorestart_with_critical_process(duthost, container_name, process_name, 
```

Comment:
Note that we currently have set a threshold in systemd which will consider the service failed if it restarts more than 3 times in 20 minutes. You will need to take this into consideration, especially when testing killing multiple processes in the container. If the container fails to restart, you will need to check if systemd has considered it failed and clear the failed condition using `systemctl reset-failed <service_name>.service`

Also, for the "extensive" mode of this test, we should test that that feature actually takes effect.

Reply:
I will update this function with consideration of the threshold and test "extensive" mode.